src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/TypesTableNode.cs

```diff
@@ -38,31 +39,34 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
             VertexHashtable typesHashtable = new VertexHashtable();
             section.Place(typesHashtable);
 
-            HashSet<TypeDesc> uniqueTypes = new HashSet<TypeDesc>();
+            ReadyToRunTableManager r2rManager = (ReadyToRunTableManager)factory.MetadataManager;
 
-            foreach (TypeDesc type in ((ReadyToRunTableManager)factory.MetadataManager).GetTypesWithAvailableTypes())
+            foreach (DefinedTypeInfo definedTypeInfo in r2rManager.GetDefinedTypes())
             {
-                int rid = 0;
-                if (type.GetTypeDefinition() is EcmaType ecmaType)
+                TypeDefinitionHandle typeDefHandle = definedTypeInfo.Handle;
+                int hashCode = 0;
+                for (; ; )
                 {
-                    if (uniqueTypes.Add(ecmaType))
+                    TypeDefinition typeDef = definedTypeInfo.Module.MetadataReader.GetTypeDefinition(typeDefHandle);
+                    string namespaceName = definedTypeInfo.Module.MetadataReader.GetString(typeDef.Namespace);
+                    string typeName = definedTypeInfo.Module.MetadataReader.GetString(typeDef.Name);
+                    hashCode ^= ReadyToRunHashCode.NameHashCode(namespaceName, typeName);
+                    if (!typeDef.Attributes.IsNested())
                     {
-                        rid = MetadataTokens.GetToken(ecmaType.Handle) & 0x00FFFFFF;
-                        Debug.Assert(rid != 0);
-
-                        int hashCode = ReadyToRunHashCode.TypeTableHashCode(ecmaType);
-                        typesHashtable.Append(unchecked((uint)hashCode), section.Place(new UnsignedConstant((uint)rid << 1)));
+                        break;
                     }
+                    typeDefHandle = typeDef.GetDeclaringType();
                 }
-                else if (type.IsArray || type.IsMdArray)
-                {
-                    // TODO: arrays in type table - should we have a recursive descent into composite types here
-                    // and e.g. add the element type to the type table in case of arrays?
-                }
-                else
-                {
-                    throw new NotImplementedException();
-                }
+                typesHashtable.Append(unchecked((uint)hashCode), section.Place(new UnsignedConstant(((uint)MetadataTokens.GetRowNumber(definedTypeInfo.Handle) << 1) | 0)));
+            }
+
+            foreach (ExportedTypeInfo exportedTypeInfo in r2rManager.GetExportedTypes())
+            {
+                ExportedType exportedType = exportedTypeInfo.Module.MetadataReader.GetExportedType(exportedTypeInfo.Handle);
+                string namespaceName = exportedTypeInfo.Module.MetadataReader.GetString(exportedType.Namespace);
+                string typeName = exportedTypeInfo.Module.MetadataReader.GetString(exportedType.Name);
+                int hashCode = ReadyToRunHashCode.NameHashCode(namespaceName, typeName);
```

Comment:
Will this work right for nested exported types?

Reply:
I don't understand how nested exported types are supposed to work as there's no "parent" in ExportedType. Are you aware of some framework assemblies using nested forwards so that I can disassemble them to see how it works? I tried to look at System.Runtime and System.Runtime.Extensions but neither seem to contain any nested forwards.

Reply:
The `Implementation` of the exported type will be another exported type if this is a nested type. It's an assembly reference otherwise. Look for e.g. the enumerators on the generic collection types. Those are nested and exported.

Reply:
In ILASM lingo:

```
.class extern forwarder System.Collections.Generic.Dictionary`2
{
  .assembly extern System.Private.CoreLib
}
.class extern Enumerator
{
  .class extern System.Collections.Generic.Dictionary`2
}
```

The difference is in the `.class` part inside the brackets.

Reply:
Fixed in 4th commit, thanks for the detailed info!